### ISSUE TITLE
Typst support

### ISSUE THREAD

**Issue by taylorh140** [terrastruct/d2#1435]
This would be great to have in Typst. They are working on using web assembly for plugins. It would be nice to have a d2 plugin.

**Comment by bugsbugsbux** [comment]
bump, dont autoclose

**Comment by dlouwers** [comment]
Hi! I'd like to work on this issue.

Based on my analysis of the D2 codebase, I believe Typst support should follow the **renderer pattern** (like `d2renderers/d2latex/`) rather than the plugin pattern, since Typst would be used for rendering diagram labels/text content, not for layout algorithms.

My proposed approach:
- Create `d2renderers/d2typst/` package
- Implement `Render()` (typst markup → SVG) and `Measure()` (dimension calculation)
- Integrate at the `d2graph` layer following the LaTeX precedent

I'm evaluating three implementation options:
1. **CLI wrapper**: Shell out to `typst compile` (simplest, requires typst installation)
2. **Rust library via CGO**: Embed typst's Rust library directly (most integrated)
3. **WASM**: Use typst's WASM build (browser-compatible, self-contained)

I'll submit a design doc for review before starting implementation. Does this approach sound reasonable to the maintainers?

cc @alixander (based on LaTeX renderer PR history)

**Comment by dlouwers** [comment]
PR submitted: #2724

Implemented Typst support using a hybrid approach:
- Native builds: CLI wrapper (requires typst installation)
- Browser builds: typst.ts WASM library (bundled, 206KB)

The implementation follows the LaTeX renderer pattern exactly and includes comprehensive tests. See the PR for full details and demo examples.

**Comment by hpr1999** [comment]
I'm pretty sure what was being asked in this issue was not to use Typst for math typesetting (which is also nice), but to have a plugin that allows _Typst_ users to create D2 diagrams within Typst.

I have successfully got Codex to whip up a solution for this, which I'm trying to massage into a proper, human-understandable solution.
If I forget to do it and someone is interested, feel free to bug me.

The sticking points were mostly related to fonts.
- Since the reading of system fonts needs filesystem access, which the Typst plugin sandbox does not provide.
- Also, D2 was measuring with its special fonts, but then, since Typst's SVG renderer does not support `@font-face`, it was rendering with the default fallbacks, which caused clipping due to differing widths. Codex got some weird hacks going. As far as I understand, it replaced the text rendering with a different code path that uses the font glyphs and renders them directly. 

Also, it is slower than the d2 watch mode, taking a full ~5 seconds when a diagram changes.

---

The current result is a working WASI build compatible with the Typst plugin-API, which can be used to write things like 

````typst
#show raw.where(lang: "d2", block: true): source => diagram(theme: 1, source, pad: 4)

```d2
A -> B
```

// or directly

#diagram("A ->")
````

inside of typst source code to have it generate a diagram.

**Comment by taylorh140** [comment]
Well I’ll look into it more. But this is very good! Thanks for your work on
this.
